src/format/mod.rs

```diff
@@ -756,6 +707,159 @@ fn format_inner<'a>(
+                bytes[4] = b'0' + (mins / 10) as u8;
+                bytes[5] = b'0' + (mins % 10) as u8;
+                &bytes[0..6]
+            }
+            Colons::Double => {
+                // colons are already in the bytes
+                let mins = off / 60 % 60;
+                bytes[4] = b'0' + (mins / 10) as u8;
+                bytes[5] = b'0' + (mins % 10) as u8;
+
+                let secs = off % 60;
+                bytes[7] = b'0' + (secs / 10) as u8;
+                bytes[8] = b'0' + (secs % 10) as u8;
+
+                &bytes[0..9]
+            }
+            Colons::Triple => &bytes[0..3],
+        };
+
+        let date = core::str::from_utf8(date).map_err(|_| fmt::Error)?;
+        result.push_str(date);
+        Ok(())
+    } else {
+        result.push('Z');
+        Ok(())
+    }
+}
+
+#[cfg(any(feature = "alloc", feature = "std", test))]
+pub(crate) fn write_rfc3339(
+    result: &mut String,
+    d: &NaiveDate,
+    t: &NaiveTime,
+    off: FixedOffset,
+) -> fmt::Result {
+    // reuse NaiveDate and NaiveTime `Debug` impls which already print ISO 8601 format.
+    // this is faster in this way.
+    d.write_into(result)?;
+    result.push('T');
+    t.write_into(result)?;
+    write_local_minus_utc(result, off, false, Colons::Single)
+}
+
+#[cfg(any(feature = "alloc", feature = "std", test))]
+pub(crate) fn write_rfc2822(
+    result: &mut String,
+    d: &NaiveDate,
+    t: &NaiveTime,
+    off: FixedOffset,
+) -> fmt::Result {
+    #[cfg(feature = "unstable-locales")]
+    let (short_months, short_weekdays) =
+        (locales::short_months(Locale::POSIX), locales::short_weekdays(Locale::POSIX));
+    #[cfg(not(feature = "unstable-locales"))]
+    let (short_months, short_weekdays) = (
+        &["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"],
+        &["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
+    );
+    write_rfc2822_inner(result, d, t, off, short_weekdays, short_months)
+}
+
+#[cfg(any(feature = "alloc", feature = "std", test))]
+fn write_rfc2822_inner(
+    result: &mut String,
+    d: &NaiveDate,
+    t: &NaiveTime,
+    off: FixedOffset,
+    short_weekdays: &[&str],
+    short_months: &[&str],
+) -> fmt::Result {
+    result.push_str(short_weekdays[d.weekday().num_days_from_sunday() as usize]);
+    result.push_str(", ");
+
+    let mut bytes: [u8; 15] = *b" 0000 00:00:00 "; // enough to fit ` YYYY HH:MM:SS `
+
+    // day of month
+    {
+        bytes[3] = b'0' + (d.day() / 10) as u8;
+        bytes[4] = b'0' + (d.day() % 10) as u8;
+        let dom = core::str::from_utf8(&bytes[3..6]).map_err(|_| fmt::Error)?;
+        result.push_str(dom);
+    }
+
+    result.push_str(short_months[d.month0() as usize]);
+
+    // year
+    {
+        // 4 digit year 1234
+        // is split into cent=12 and subcent = 34
+        let cent = d.year() / 100;
+        let subcent = d.year() % 100;
+        bytes[1] = b'0' + (cent / 10) as u8;
```

Comment:
potentially some of this can be shared with `NaiveDate::write_into`

Reply:
Not exactly sure what you had in mind, but I added a `write_double_digit` method which makes this clearer